crates/bevy_ecs/src/change_detection.rs

```diff
@@ -0,0 +1,152 @@
+use crate::component::{Component, ComponentTicks};
+use bevy_reflect::Reflect;
+use std::ops::{Deref, DerefMut};
+
+pub trait ChangeDetectable {
+    /// Returns true if (and only if) this value been added since the last execution of this
+    /// system.
+    fn is_added(&self) -> bool;
+
+    /// Returns true if (and only if) this value been changed since the last execution of this
+    /// system.
+    fn is_changed(&self) -> bool;
+
+    /// Manually flags this value as having been changed. This normally isn't
+    /// required because accessing this pointer mutably automatically flags this
+    /// value as "changed".
+    ///
+    /// **Note**: This operation is irreversible.
+    fn set_changed(&mut self);
+}
+
+macro_rules! change_detection_impl {
+    ($name:ident < $( $generics:tt ),+ >, $target:ty, $($traits:ident),*) => {
+        impl<$($generics),*: $($traits),*> ChangeDetectable for $name<$($generics),*>
```

Comment:
This needs to be
```suggestion
        impl<$($generics),*: $($traits)+*> ChangeDetectable for $name<$($generics),*>
```
or if you want to only allow at most single trait:
```suggestion
        impl<$($generics),*: $($traits)?> ChangeDetectable for $name<$($generics),*>
```

Reply:
Sadly I don't think `impl<$($generics),*: $($traits)+*>` is allowed since `+` is not a valid separator in `macro_rules!` variadics.
So in the case where this would require more than one trait, we would have to switch to a proc macro.
However, this only needs to support a single trait, so it's probably best to use `?` to better express my intent.